### ISSUE TITLE
Range Scoping Doesn't Support Empty Ranges

### ISSUE THREAD

**Issue by ElizabethHudnott** [EMGroup/js-eden#209]
Example:

l = [ ];
j is -1 if i == 1 else l[i] + 1;
l2 is j with i is 1..l#;

l2 is evaluated as [-1] but should be [ ] according to the laws of math.

**Comment by knicos** [comment]
Not sure how this is "laws of math" rather than standard loop interpretation. What about doing 1..-40? Seems like as valid an option as doing 1..0 which should actually generate two results not 1. Ranges are not exclusively used as list indices.

**Comment by knicos** [comment]
Fixed to allow reverse ranges. There can never be any empty range.

**Comment by ElizabethHudnott** [comment]
So, in your first reply, you're saying

`x with i is 1..-40`

has two values, one when i equals 1 and one when it equals -40.  I don't follow the logic.  How does "From 1 to -40 counting up in 1s" end up as two values?  Arguably it could result in an infinite number of values, but that's not a practical result to compute.  The laws of math definition is that we're talking about the set:

`{x | 1 <= i <= -40}`

which evaluates to the empty set (though obviously with retains ordering and duplicate values, returning a list rather than a set).  This is how with is described in the PPIG paper.

In terms of a loop based interpretation, I guess:

`i = 1;
while (i <= -40) {
  ...compute value...
  i++;
}`

Why would the range be reversed?  Surely that would only happen when the increment is a negative number?  But the increment is 1 by default.  What do you envisage happening if the increment is explicitly specified as +1?

When the upper (or lower) bound is determined by an observable then I think it's quite common to have/want an empty list returned when the lower bound exceeds the upper bound.  People did this quite often when they used to write functions and for loops.

This is the behaviour Meurig was expecting yesterday, when he wrote in his e-mail:

> I noted in passing that if you use a range definition of the form
> `changeGiven is changeGiven_ix with ix is 1..coinlist#;`
> which according to normal mathematical convention would yield the empty list, you get some anomalous value. For this reason, I have used an alternative definition viz:
> `changeGiven is [] if coinlist#==0 else changeGiven_ix with ix is 1..coinlist#;`

**Comment by knicos** [comment]
What "law of maths" is that? Doesn't seem like a law so much as a programming convention being incorrectly brought over. Please read my comment more carefully, I say that 1..0 has two values, not 1..-40.

The increment can easily be switched from 1 to -1 automatically depending on range direction.

If the increment is explicitly specified as the opposite sign to that required a runtime error is reported to the script saying "Infinite Range Detected".

The direction and error checking is done at runtime so it does not matter if the upper and lower are expressions.

Regarding Meurigs expected result ... again a miss understanding of what the range is due to expectations being brought over from "for" loops. It is NOT a "for" loop but, for example, `x is 1..5` is equivalent to `x in [1,2,3,4,5]`.

**Comment by ElizabethHudnott** [comment]
The mathematical convention is that `x in 1..5` is usually shorthand for the set comprehension  `x in { i | 1 <= i <= 5}`.  E.g. Z Notation.

**Comment by knicos** [comment]
We are dealing with ordered lists, not sets!

**Comment by knicos** [comment]
As a compromise, if you really need an empty result I could allow the following to generate an empty list instead of an error:

`x is 1..1..0;`

Due to it forcing a positive increment.

**Comment by ElizabethHudnott** [comment]
I think the sets example generalizes to lists in a trivial way (in observational terms) that faithfully reflects how `..` is used with sets in other notations (including regular pencil and paper math).

(In logic terms: the generalization is by adding the criteria that every member in the set is a value in the corresponding list and vice versa, and that the list is in ascending order).

What does x..y correspond to in terms of basic Boolean logic, etc. in your interpretation?  It must surely have the additional complication that explicitly spells out that the list either in ascending order or descending order.

**Comment by knicos** [comment]
For me it is intuitive that 5..1 is a descending order list. Any boolean logic is already complicated by arbitrary (including negative) increments that can be specified.

There are three questions:
1) How likely are descending order lists vs empty range requirements.
2) What do school children and non-programmers think about, say, 5..1. This seems like an obvious thing to write and yet it would generate any empty list.
3) Are we going to tie into mathematical convention even though we a not aiming EM at mathematicians?

**Comment by ElizabethHudnott** [comment]
Well, the latest thinking is that one of the main objectives of CONSTRUIT! going forward is to create OERs to support the Computing At Schools curriculum (additionally with a learning curve towards experimenting with simple modifications).  With CaS "programmer", "mathematician" and "school children" are converging towards the same thing.  Meurig is even creating a construal about sets.

My additional concern is that when the upper or lower bound is an observable reference, and that observable is gradually adjusted (e.g. using a slider) then the list gradually shrinks towards a minima, but under the automatic reversal of bounds thing, trying to go smaller still results in the list surprisingly "bouncing back" and growing again.

Basically, sure, 5..1 naturally suggests a descending order (perhaps).  But does 1..x?  or x..y?  This kind of intuition only seems to work when both bounds are constants.

**Comment by knicos** [comment]
Then my compromise should be fine? `x..1..y`. So that the constant version remains intuitive and the variable version becomes explicit to avoid misunderstandings.

?

**Comment by knicos** [comment]
Also, the slider bounce example ... it would be fine with me if sliding further actually resulted in something happening rather than just staying as empty regardless of any further sliding.

**Comment by knicos** [comment]
Alright. I've made it fit convention, will commit later. It will now generate empty results.
